### DISCUSSION TITLE
-

### DISCUSSION THREAD

**Discussion by DavidArno** [dotnet/runtime#80293] [General] [answered]
-

**Comment by CyrusNajmabadi** [comment, marked as answer]
Moving to runtime. This is not a lang issue.

**Reply by DavidArno** [reply]
As "helpful" as ever, eh Cyrus. I deliberately put it on csharplang as it was intended as a language-centric discussion. But whatever. 🤦

**Reply by CyrusNajmabadi** [reply]
> a language-centric discussion. 

This is a runtime issue.  The language will pickup and support whatever enumeration policy any particular type wants to expose.  For the type System.Range, the BCL owns the enumeration strategy (or lack thereof).  The original post contained absolutely nothing wrt to the rules of the language, and this idea has already been assessed in the past to belong to the runtime.  So this was moved appropriately.  If the runtime gets enough feedback that this is desirable, it can be something that may be considered here.

**Comment by CyrusNajmabadi** [comment]
Original topic:

```c#
foreach (var x in 1 .. 10)
{
    Console.WriteLine(x);
}
```

Try and write the above code and one gets:

```
error CS1579: foreach statement cannot operate on variables of type 'Range' because 'Range' does not 
contain a public instance or extension definition for 'GetEnumerator'
```

Many have asked, why doesn't Range implement IEnumerable<int> and the reason given - as I understand it - is because of "from end" ranges. What is the following supposed to do?

```c#
foreach (var x in ^1 .. ^10)
{
    Console.WriteLine(x);
}
```

So because of this, Range doesn't implement IEnumerable<int>.

Now I'll concur that the ^1 at the start of the range doesn't make much sense. But that ^10 is actually useful. Consider what we might expect the output of this code to be:

```c#
foreach (var x in 1 .. 10)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```

I'm assuming most people will be like me here and expect `1 2 3 4 5 6 7 8 9`, as if the end of the range is a "from beginning" value, then it's a non-inclusive range. As "from end" ranges already have the concept of being inclusive ranges, we actually have a purpose for them here:

```c#
foreach (var x in 1 .. ^10)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```

would output `1 2 3 4 5 6 7 8 9 10`.

And of course we can easily do reverse ranges too:

```c#
# Outputs 10 9 8 7 6 5 4 3 2
foreach (var x in 10 .. 1)
{
    Console.Write($"{x} ");
}
Console.WriteLine(); 

# Outputs 10 9 8 7 6 5 4 3 2 1
foreach (var x in 10 .. ^1)
{
    Console.Write($"{x} ");
}
Console.WriteLine();
```

To my mind, the usefulness of being able to so succinctly express incremental and decremental ranges, that can be either inclusive or exclusive, far outweighs the slight stumbling block of "what does ^1 .. 10 mean?" Have a warning that it's meaningless and ignore it seems the simplest solution to that one.

**Comment by CyrusNajmabadi** [comment]
I'm locking this as the OP deleted their post.  That indicates no more desire to discuss things afaict.
